[RFC] rewrite of the man plugin

nhooyr opened the pull request:
For details see https://github.com/nhooyr/neoman.vim/pull/15 and https://github.com/neovim/neovim/pull/4168#issuecomment-195880774

I just need to write the new documentation and a little section on top of the files for the maintainer and date last changed.

nhooyr committed:
man.vim: rewrite

- Smart autocomplete. It's automatically sorted, filtered for duplicates
  and even formats the candidates based on what is needed. For example,
  `:Man 1 printf<TAB>` will show the pages that are in section 1m as
  'page(sect)' to let you know they are in a more specific section.
- Instead of trying to unset $MANPAGER we use the -P flag to set the
  pager to cat
- Always use the section arg '-s', it makes the code much simpler
  (see comment in s:man-args).
- A manpage name starting with '-' is invalid. It's fine for sections
  because of the use of '-s'.

- The tagstack is an actual stack now, makes it much simpler.
- By using v:count and v:count1, the plugin can explicitly check whether
  the user set a count, instead of relying on a default value (0) that
  is actually a real manpage section.
- Extraction of a manpage reference is much more simple. No giant long
  complicated regexes. Now, the plugin lets `man` handle the actual
  validation. We merely extract the section and page. Syntax regexes are
  a bit more specific though to prevent highlighting everything.
- Multilingual support in the syntax file. Removed the cruft that was only
  relevent to vim. Also simplified and improved many of the regexes.

- Using shellescape when sending the page and sect as arguments
- In general, the code flow is much more obvious.
- man#get_page has been split up into smaller functions with explicit
  responsibilties
- ':help' behavior in opening splits and manpages
- Comments explaining anything that needs explaining and isn't
  immediately obvious.
- If a manpage has already been loaded but if it were to reloaded at the
  current width which is the same as the width at which it was loaded at
  previously, it is not reloaded.

- Use substitute to remove the backspaced instead of `col -b`, as the
  latter doesn't work with other languages.
- Open paths to manpages
- It uses cWORD instead of cword to get the manpage under the cursor, this
  helps with files that do not have (,) in iskeyword. It also means the
  plugin does not set iskeyword locally anymore.
- <Plug>(Man) mapping for easy remapping
- Switched to single quotes wherever possible.
- Updated docs in $VIMRUNTIME/doc/filetype.txt (still need to update
  user-manual)

- Always call tolower on section name. See comment in
  s:extract_page_and_sect_fpage
- Formatting/consistency cleanup
- Automatically map q to ':q<CR>' when invoked as $MANPAGER
- It also fully supports being used as $MANPAGER. Setting the name and
  stuff automatically.
- Split up the setlocals into multiple lines for easier readability
- Better detection of errors by redirecting stderr to /dev/null. If an
  error occured, stdout will be empty.

- Functions return [sect, page] not [page, sect]. Makes more sense with
  how man takes the arguments as sect and then page.
- Pretty prints errors on a single line.
- If no section is given, automatically finds the correct section for
  the buffer name. It also gets the correct page. See the comment in
  s:get_page
- If $MANWIDTH is not set, do not assign directly to $MANWIDTH because
  then $MANWIDTH will always stay set to the same value as we only use
  winwidth(0) when the global $MANWIDTH is empty. Instead we set it
  locally for the command.
- Maintainer notes on all files.

justinmk commented on runtime/plugin/man.vim:
```diff
@@ -1,6 +1,21 @@
+command! -bang -complete=customlist,man#Complete -nargs=* Vman call
+      \ man#get_page(<bang>0, 'vsplit', <f-args>)
+command! -bang -complete=customlist,man#Complete -nargs=* Tman call
+      \ man#get_page(<bang>0, 'tabe', <f-args>)
```
Adding new commands should not be done lightly, especially for a default plugin. E.g. if user has `:Tmux` command, now tab-complete will complete `Tman` first. This drives people crazy.

Let's hold off on the new commands and leave the question of how to provide these alternate modes for another PR.

ZyX-I replied:
Of all these commands only `:Man` is definitely needed. `:Sman` may be needed to avoid side-effects of running `:split` before `:Man` (AFAIK you will at least have a few \*Win\* autocommands with unexpected patterns), but others should be possible to emulate using modifiers like `:vertical` or `:tab`.

nhooyr replied:
yup, I'll remove them.

@justinmk should we make the behaviour be exactly like `:help`? Horizontal split by default?

justinmk replied:
No, behavior should remain `:tabnew` by default. That is the best approximation we have for "pager" behavior which preserves window layout, for the time being.

I suggest a slight change, however:

    :execute (tabpagenr()-1).'tabnew'

This opens tab _before_ current tab, so that `q` returns to the original tab. Again, to approximate traditional behavior of `:!man`.

Also let's keep any new commands for a separate PR. I want to move quickly on this one ;)

nhooyr replied:
Alright, will change it.

justinmk commented on runtime/plugin/man.vim:
```diff
@@ -1,6 +1,21 @@
 
-command! -count=0 -nargs=+ Man call man#get_page(<count>, <f-args>)
+if !exists("g:find_man_window")
+  let g:find_man_window = 1
```
Other than the unfortunate `g:loaded_` convention, plugins must always prefix globals with the plugin name: `g:man_...`

nhooyr replied:
changed.

justinmk commented on runtime/syntax/man.vim:
```diff
@@ -1,67 +1,34 @@
-  runtime! syntax/ctrlh.vim
-else
-  source <sfile>:p:h/ctrlh.vim
-endif
```
I don't know what this was for, but can you explain why it's not needed anymore?

nhooyr replied:
Its useless as far as I can tell. If you remove line 122 `execute "silent! keepjumps %substitute,.\b,,g" of `autoload/man.vim` you'll see what it does. Man outputs backspaced text but in order to display it properly, we need to remove all those backspaces. This syntax file highlights them, whereas the substitution removes them. Thus if the substitution removes them, there is no need for the syntax file to be highlighting them. No idea why it was there before, man used to be piped into `col -b` which does the same thing, but doesn't work with other languages.

justinmk replied:
:+1: Thanks. Consider mentioning that in the commit message.

justinmk commented on runtime/syntax/man.vim:
```diff
@@ -1,67 +1,34 @@
+highlight default link manSubHeading     Function
+
+if getline(1) =~# '^\f\+([23])'
+  syntax include @cCode syntax/c.vim
```
I'm not sure, but will this load user-installed `syntax/c.vim`? That might be an unintended source of problems. Before, it was explicit:


    syntax include @cCode <sfile>:p:h/c.vim

nhooyr replied:
It will, I'll change it back.

justinmk commented on runtime/syntax/man.vim:
```diff
@@ -1,67 +1,34 @@
+syntax match manLongOptionDesc  "^\s\+--[a-z0-9]\S*"
+" prevent manSectionHeading from matching last line
+let s:l = line('$')
+execute 'syntax match manSectionHeading  "^\%(\%>1l\%<'.s:l.'l\)\%(\S.*\)\=\S$"'
```
Why not use `line('$')` directly in the `execute` line? This is very unusual. If it's for performance, forget it.

nhooyr replied:
done

justinmk commented on runtime/syntax/man.vim:
```diff
@@ -1,67 +1,34 @@
+  " skip first heading
+  keepjumps call search('^\%(\S.*\)\=\S$')
+  " get second heading, aka the synopsis heading
+  let l = escape(getline(search('^\%(\S.*\)\=\S$', 'n')), '\')
```
This creates or overwrites a global `g:l`. Put this if-block in a private function.

nhooyr replied:
I'm gonna just change the `l` to `s:l`.

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -1,37 +1,26 @@
-  endif
+  nnoremap <silent> <buffer> <C-]>    :call man#get_page(g:find_man_window, 'edit')<CR>
+  nnoremap <silent> <buffer> <C-t>    :call man#pop_tag_stack()<CR>
+  nnoremap <silent> <buffer> q :q<CR>
```
This is not the same as `<C-W>c`. The old mapping was intentional. Can't change this.

nhooyr replied:
will change it back.

noahfrederick commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,211 @@
+  silent execute a:cmd 'man://'.a:page.(empty(a:sect)?'':'('.a:sect.')')
+  setlocal modifiable
+  " remove all the text, incase we already loaded the manpage before
+  silent keepjumps normal! gg"_dG
```
Why not `silent keepjumps %delete _` ?

nhooyr replied:
changed.

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -1,37 +1,26 @@
-" when doing ctrl-[ on a man page reference.
-if exists('$MANPAGER')
-  let $MANPAGER = ''
-endif
```
why was this removed?

nhooyr replied:
I'll add this back, more info here on why its necessary http://comments.gmane.org/gmane.editors.vim.devel/29085

justinmk replied:
General theme is "don't remove things without knowing why they exist". 

If there are other cases where things were removed without knowing, please revert those changes, or put them in a separate PR.

nhooyr replied:
turns out, this doesn't even work properly, on OSX it screws everything up, because of `http://permalink.gmane.org/gmane.editors.vim.devel/29629`. Instead its better to make the shell unset the variable, and then run the man command.

justinmk replied:
Yes, the problem is known, but we can't just remove it. It was discussed here:

https://gitter.im/neovim/neovim?at=5645154210f6aea94fdd7f08

nhooyr replied:
Not gonna remove it, just make it better. I'll push all my changes in a minute.

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -1,37 +1,26 @@
-setlocal iskeyword+=\.,-,(,)
-
-setlocal buftype=nofile noswapfile
-setlocal nomodifiable readonly bufhidden=hide nobuflisted tabstop=8
```
Please restore this line. Avoid needless changes like this, it makes the diff bigger and harder to review.

nhooyr replied:
@justinmk I find it much more readable the way I have it formatted as a nice list.

noahfrederick commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,211 @@
     silent keepjumps $delete _
   endwhile
-  setlocal nomodified
+  keepjumps normal! gg
```
```vim
keepjumps 1
```

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,211 @@
-try
-  if !has('win32') && $OSTYPE !~? 'cygwin\|linux' && system('uname -s') =~? 'SunOS' && system('uname -r') =~? '^5'
-    let s:man_sect_arg = '-s'
-    let s:man_find_arg = '-l'
```
Different systems have different flags. How does the new code handle this?

nhooyr replied:
When I was working on neoman, since I couldn't test any of this, I decided to remove for simplicity's sake, but I'll add it back in.

noahfrederick commented on runtime/syntax/man.vim:
```diff
@@ -1,67 +1,34 @@
+  keepjumps call search('^\%(\S.*\)\=\S$')
+  " get second heading, aka the synopsis heading
+  let l = escape(getline(search('^\%(\S.*\)\=\S$', 'n')), '\')
+  keepjumps normal! gg
```
Here too:

```vim
keepjumps 1
```

noahfrederick commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,211 @@
-    return s:man_sect_arg.' '.a:sect.' '.a:page
+    for i in range(len(candidates))
+      let candidates[i] = substitute((fnamemodify(candidates[i], ":t")), find, repl, "")
+    endfor
```
Probably more performant:

```vim
call map(candidates, 'substitute((fnamemodify(v:val, ":t")), '.find.', '.repl.', "")')
```

nhooyr replied:
changed.

noahfrederick commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,211 @@
+    let fpage = a:000[0]
+  else
+    let fpage = a:000[1].'('.a:000[0].')'
   endif
```
This block is harder to follow than what you replaced. Why the change from `a:0` to `len(a:000)`? Why not rely on `get()`? Restoring and updating the documentation of the function parameters above would be good.

nhooyr replied:
My mistake, I forgot about `a:0`. I don't see how I can use `get()` here, since I only add the parenthesis if the section was given as well. My parsing function works a bit differently, it only takes the full page ('page(sect)' or just `page`) instead of two separate variables.

I'll add some comments to explain the code though.

noahfrederick commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,211 @@
-  setlocal colorcolumn=0 foldcolumn=0 nonumber
-  setlocal nolist norelativenumber nofoldenable
+function! s:error(msg) abort
+  redraws!
```
Avoid abbreviation:

```vim
redrawstatus!
```

nhooyr replied:
wow, I didn't even know that was an abbreviation lol. I've never seen its full form.

nhooyr commented:
Alright, I updated it based on feedback and rewrote the commit messages.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -1,6 +1,11 @@
-let s:man_tag_depth = 0
+" Ensure Vim is not recursively invoked (man-db does this)
+" More info here http://comments.gmane.org/gmane.editors.vim.devel/29085
+let s:man_cmd = 'unset MANPAGER; man 2>/dev/null '
```
@justinmk instead of the previous code where `$MANPAGER` was emptied, this actually removes it from the environment.

noahfrederick replied:
Note that `unset MANPAGER; man 2>/dev/null` will fail for `set shell=fish` (which Vim claims to support in some capacity—not sure if that extends to default plugins). Fish syntax would be `set -e MANPAGER; man ^/dev/null`.

nhooyr replied:
I'll add a simple if statement.

justinmk replied:
Can we please wait on this change until https://github.com/neovim/neovim/issues/3265 is implemented?

nhooyr replied:
@justinmk we could leave it like this, and then change it when #3265 is implemented. Its a few lines, nothing major.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +16,217 @@ catch /E145:/
+  let find = '\(.\+\)\.\%('.s:man_extensions.'\)\@!\'
+  " if the page is a path, complete files
+  if a:sect ==# '*' && a:page =~# '\/'
+    "TODO why does this complete the last one automatically
```
@justinmk @noahfrederick can you guys test this? Try `:Man ./<TAB>`. For some reason, it completes the last result automatically for me, yet this doesn't happen with `:Man <TAB>`?

nhooyr commented on runtime/syntax/man.vim:
```diff
@@ -1,67 +1,39 @@
+  " get second heading, aka the synopsis heading
+  let s:l = escape(getline('.'), '\')
+  " why bsd...
+  if s:l ==# 'LIBRARY'
```
For some reason FreeBSD has a extra section between the NAME and SYNOPSIS which means I can't just always select the second section for the c highlighting...

https://www.freebsd.org/cgi/man.cgi?query=printf&apropos=0&sektion=3&manpath=FreeBSD+10.2-RELEASE&arch=default&format=html

I've added a special case, but we all hate special cases. I was thinking of a variable like `g:man_synopsis` , which contains a regex used to match the synopsis. So by default it would probably be `^SYNOPSIS$` but you could change it to add in other languages yourself.

@justinmk what do you think?

justinmk replied:
I don't know what this is for. Can it wait in another PR? We need to keep the default plugins reliable and simple. I'd like to keep this PR focused on just fixing the major problems you noticed, in a cross-platform manner. New features like tab-complete which are a big win (without tons of code or bugs) are also welcome. But main objective is to address low-hanging fruit.

nhooyr replied:
@justinmk for manpages with c code in them. e.g. https://www.freebsd.org/cgi/man.cgi?query=printf&apropos=0&sektion=3&manpath=FreeBSD+10.2-RELEASE&arch=default&format=html

See the synopsis section with the c code? The manpage plugin currently matches `SYNOPSIS` to start the c highlighting, except this obviously won't work in other languages. Thus above, I'm matching the second header in manpages of sections 2/3, which is usually `SYNOPSIS` but FreeBSD decided to make it `LIBRARY` and the third header `SYNOPSIS`.

Above I've added a special case for FreeBSD but the second option is a user interface, a variable the user can use to define a regex for where the `SYNOPSIS` begins in their language.

But its not a big issue, we can address it in a later P.R.

justinmk replied:
Got it, thanks. The C highlighting seems to work well, in the current plugin (before this PR), on linux and OSX. As long as that doesn't regress, I suppose it wouldn't be bad to try to cover more cases... but separate PR would be best.

nhooyr replied:
I'm gonna revert this, the c highlighting does regress, there are too many special cases to this approach of making the second heading always the synopsis heading.

justinmk commented on runtime/doc/filetype.txt:
```diff
@@ -512,30 +512,42 @@ Local mappings:
+			are we to tell where the C code begins? This variable is a regex
+			that is used to match the SYNOPSIS heading.
+			Modify it for the language(s) you will be using.
+			Defaults to '\V\^SYNOPSIS\$'.
```
This is unnecessary configuration and the amount of documentation required to describe it is disproportionate to its importance. It's trivial to extend the regex by adding extra words from other languages. If we want to support other languages by doing that, great. If new languages are to be supported, add more equivalents to the regex. But don't expose an option. The 5 people that actually take the time to discover this option could just as easily send a 1-line PR so that everyone viewing manpages in that language will benefit.

nhooyr replied:
Alright, I'll change it and add as many languages as I can.

justinmk replied:
How about just the top 5 or 10 languages--assuming you can find manpages in those languages.

nhooyr replied:
Yup, I found 6. We can add more in the future, doesn't really matter.

justinmk commented on runtime/doc/filetype.txt:
```diff
@@ -512,30 +512,42 @@ Local mappings:
+			Defaults to '\V\^SYNOPSIS\$'.
+g:no_man_maps
+			If set, no mappings are created in man buffers.
+			Defaults to 0.
```
`no_plugin_maps` is another ancient tradition like `loaded_foo`. Plugin-specific options must be prefixed with the plugin name.

justinmk commented on runtime/doc/filetype.txt:
```diff
@@ -512,30 +512,42 @@ Local mappings:
+Man			Display the manual page for <cWORD> in a new tab.
+Man {name}		Display the manual page for {name} in a new tab.
+Man {sect} {name}	Same as above except specify the section.
+Man {name}({sect})	Alternate syntax to above.
```
`man` doesn't support this syntax, so why should `:Man`? This seems like feature bloat.

nhooyr replied:
I'll remove this. I also noticed a few problems in autoload/man.vim. I'm gonna refactor it a bit.

nhooyr replied:
Oh never mind, I was mistaken, it was fully intended. It lets you see which section the manpage is on. E.g. `:Nman print<TAB>` shows `:Nman printf(3)`. You know exactly which section you will be taken to. Its either like that or, it complete it like `:Nman 3 printf` which just looks weird when you have a long list of completions.

justinmk replied:
Ok. Completion is a fair use case, I guess.

nhooyr replied:
I did notice other problems in autoload/man.vim that I'm gonna fix.

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -1,37 +1,27 @@
+  nnoremap <silent> <buffer> <C-t>    :call man#pop_tag()<CR>
   nnoremap <silent> <nowait><buffer>  q <C-W>c
-  if &keywordprg !=# ':Man'
-    nnoremap <silent> <buffer> K      :call man#get_page(v:count, expand('<cword>'))<CR>
```
It looks like you removed the count feature, so `[count]K` and `[count]<C-]>` won't work anymore. That needs to be added back. The "bang" feature should wait until another PR.

nhooyr replied:
Will do. Are you sure about the bang? its very simple, its a 1 line of code in autoload/man.vim and 3 lines in plugin/man.vim.

justinmk replied:
`g:man_find_window` is sufficient. Making it toggle on-the-fly with bang is unusual and is not in demand. These experimental features should be left to external plugins.

nhooyr replied:
I understand that, but it isn't even a new line of code. Its just for convenience. Sometimes I want to go to the closest man window, and sometimes I don't. Its really easy to just use the bang to switch between the two behaviours. That is the whole purpose of the bang, to flick easily between two separate behaviours.

justinmk replied:
Can't afford to discuss every detail of this PR. Too many lines changed. That's why I requested focused changes. I didn't want to debate an entirely new plugin.

Default plugins must be conservative. It would really help if the non-critical features were left out of this PR.

nhooyr replied:
I'll remove it. Is the count feature really worth it, it won't even work for all man page sections like `n` and `0P`. And I'm getting very weird behaviour with the man plugin and using the count. Random errors about it not being able to parse random text as it selects the wrong text underneath the cursor. Or when a error happens, its printed as many times as the count lol. I really don't think we should implement it.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -1,6 +1,16 @@
+  let s:man_cmd = 'set -e MANPAGER; man ^/dev/null'
+else
+  let s:man_cmd = 'unset MANPAGER; man 2>/dev/null '
+endif
```
When `unlet $FOO` is implemented, we must remember to move this back to the ftplugin file...

nhooyr replied:
absolutely, I'll add a TODO.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +21,216 @@ catch /E145:/
-  setlocal nolist norelativenumber nofoldenable
+function! s:error(msg) abort
+  redrawstatus!
+  echon 'man: '
```
prefix should be `man.vim:`

nhooyr replied:
got it.

nhooyr commented:
@justinmk Ok so I've implemented everything as I think you want it.

I have several questions though. 

First, can I replace all the `substitute()`s with something else that is perhaps a bit more suited? See line 112 in `autoload/man.vim`.

Second, is deleting the bottom lines necessary? I think only deleting the first line is necessary when using r! to read manpage in. See line 137 in `autoload/man.vim`.

Third, I found a bug in both neovim and vim's completion mechanism. See nhooyr/neoman.vim#17 and line 205 in `autoload/man.vim`. Should I post it onto the vim mailing list first or should I create a neovim issue?

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +22,207 @@ catch /E145:/
+  if fnamemodify(tail, ':e') =~# s:man_extensions
+    let tail = fnamemodify(tail, ':r')
+  endif
+  " TODO(nhooyr) all substitutes to matchlist/matchstr maybe?
```
Can I replace all the `substitute()`s with something else that is perhaps a bit more suited?

tweekmonster replied:
Since you're only capturing one group, take a look at `matchstr()`.  `matchstr(tail, '\f\+')` should be good enough for `page`.  You can `\zs` and to extract the the text for `sect`.

nhooyr replied:
fixed

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +22,207 @@ catch /E145:/
   endwhile
+  " TODO(nhooyr) is deleting the bottom lines necessary?
+  " I think only deleting the first line is necessary when using r! to
+  " read manpage in.
```
Is deleting the bottom lines necessary? I think only deleting the first line is necessary when using r! to read manpage in.

justinmk replied:
Don't know. If it doesn't hurt anything, I would say leave it as it was. Up to you though.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +22,207 @@ catch /E145:/
+  let find = '\(.\+\)\.\%('.s:man_extensions.'\)\@!\'
+  " if the page is a path, complete files
+  if empty(a:sect) && a:page =~# '\/'
+    "TODO(nhooyr) why does this complete the last one automatically
```
I found a bug in both neovim and vim's completion mechanism. See nhooyr/neoman.vim#17. Should I post it onto the vim mailing list first or should I create a neovim issue?

nhooyr commented:
Found a small bug in the completion, fixing it

tweekmonster commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +22,207 @@ catch /E145:/
-    exec "unlet s:man_tag_lin_".s:man_tag_depth
-    exec "unlet s:man_tag_col_".s:man_tag_depth
-    unlet s:man_tag_buf s:man_tag_lin s:man_tag_col
+function! man#complete(ArgLead, CmdLine, CursorPos) abort
```
It looks like you broke consistency and started capitalizing variables.  It's a little weird.

nhooyr replied:
Its the standard vim completion signature so I just went with it.

tweekmonster replied:
I wouldn't call it standard.  Only found two instances of those arguments.  One in `map.txt` and one in `sql.vim`.  IMO, source consistency trumps standards if it doesn't stop code execution.  I'm not that hung up on it though 🙂

nhooyr replied:
fixed

tweekmonster commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +22,207 @@ catch /E145:/
-function s:find_page(sect, page) abort
-  let where = system('/usr/bin/man '.s:man_find_arg.' '.s:cmd(a:sect, a:page))
-  return (where =~# '^ */')
+function! s:MANDIRS() abort
```
Same comment about consistency.  Also, if you don't plan to use this function elsewhere, it might be better to put the body in the `s:get_candidates()` function since it's pretty short.

nhooyr replied:
fixed

tweekmonster commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +22,207 @@ catch /E145:/
+
+function! man#normalize_page()
+  " remove all those backspaces
+  execute "silent keepjumps %substitute,.\b,,g"
```
I'm getting `E486: Pattern not found: .^H`.  I think this needs the `e` flag as well.

nhooyr replied:
fixed

nhooyr commented on runtime/plugin/man.vim:
```diff
@@ -1,6 +1,11 @@
-command! -count=0 -nargs=+ Man call man#get_page(<count>, <f-args>)
+let g:man_find_window = get(g:, 'man_find_window', 1 )
+
+command! -count=10 -complete=customlist,man#complete -nargs=* Man call
```
In `man.vim` count is set to 0 and used as a extra argument. Two problems with this approach. First, because count is only a number you can never use a letter in the section names. `Man 1ssl cms` will throw an error on OSX even though it is a valid manpage because 3 extra arguments were sent to man.vim: `Man 0 1ssl cms`. If you use a number section it looks like `Man 3 printf`. 

Second, this default value of 0 on count is used to indicate an empty section. On some systems 0 is actually a manpage (https://en.wikipedia.org/wiki/Man_page#Manual_sections) so you can never actually specify manpage 0 because it will always be interpreted as empty.

To solve the first problem I made count a separate parameter. To solve the second problem I've made the default value of count 10 but imo that's an ugly solution. Anyone have a better idea?

tweekmonster commented:
@nhooyr I think the man pages could have better highlighting if you take advantage of the `^H` and `_` that's inserted between characters.  Not sure if you were aware of this, but those characters were how old printers printed bold and underlined text.  The `^H` would cause the print head to back up so that the next character would stamp over the previous.  So, `X^HX` would stamp `X` twice bolding it, and `_^HX` would stamp `X` over the underscore to underline it.

You could wrap anything that was meant to be bold with `*` using: 

```vim
%s/\%(\(.\)\b\1\)\+/\='*'.substitute(submatch(0), '\(.\)\b\1', '\1', 'g').'*'/g
```

It would turn `H^HHi^Hi` into `*Hi*`

Likewise for underlines using `|` pipes:

```vim
%s/\%(_\b.\)\+/\='|'.substitute(submatch(0), '_\b\(.\)', '\1', 'g').'|'/g
```
`man` on OS X seems to do it by default if it's being piped.  On Linux (I only tested Ubuntu), it prints plain text when piped, which is why I got the error about the substitution failing.  It can be forced with `man -Tutf8`, but I'm not sure if that's portable for all distros.

nhooyr commented:
Interesting, I was not aware. I'll take a stab at it tomorrow. Good night!

mhinz commented on runtime/ftplugin/man.vim:
```diff
@@ -1,41 +1,39 @@
+if !exists('g:no_plugin_maps') && !exists('g:no_man_maps')
+  nnoremap <silent> <buffer> <C-]> K
+  nnoremap <silent> <buffer> <C-t>    :call man#pop_tag()<CR>
+  nnoremap <silent> <nowait><buffer>  <C-W>c
```
This leads to a "No mapping found" message since there is no RHS.

nhooyr replied:
fixed

nhooyr commented:
I think I'll add the suggestion made by @tweekmonster in another PR. This looks good for now.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +20,197 @@ catch /E145:/
-    if &filetype !=# 'man'
-      let editcmd = 'tabnew'
+" find the closest man window above/left
+" TODO(nhooyr) more ways to open manpages
```
Need to figure out a way to also allow opening a manpage in the current window/vertically/horizontally.

nhooyr commented:
@tweekmonster I think thats what `syntax/ctrlh.vim` does. I might be able to add that into this PR.

tweekmonster commented:
Wow good find.  Didn't even know that existed.  But, it seems that it only adds highlighting.  It'll still be better to run the substitutions so that `<cword>`, cursor movement, and selections aren't hindered.

Adding delimiters to the text and a syntax group for them might be better, but may also complicate your existing syntax patterns.  If you post-process the text, you could use `matchaddpos()` to mark the bold/underline text and forego adding delimiters.

With that said, don't block the PR for this feature if you think it'll change too much of the existing script.  Deal with the low hanging fruit first.

nhooyr commented:
@tweekmonster 

It actually used to be part of `man.vim`. See my discussion with @justinmk above. 

I think I'll wait until this is merged and then look into that. I'm not too familiar with how syntax highlighting works so I'm gonna have to read `:h syntax.txt` before I start implementing that. You can take a stab at it if you want. Otherwise I think this PR is practically finished. Just need an answer to my comment on having the default count at 10.

nhooyr commented:
I think it is ready to be merged. I do not think there is a better solution for a default count and I fixed the TODO for deleting lines on top. 

What might be coming in other PRs:

- backspaced syntax highlighting
- how to handle splitting in multiple directions
- redisplaying buffer instead of reloading manpage
- a fix for the autocomplete bug (still gotta report it to vim)

justinmk commented on runtime/ftplugin/man.vim:
```diff
@@ -4,8 +4,10 @@ endif
-  call man#normalize_page()
-  silent execute 'file '.'man://'.tolower(matchstr(getline(1), '^\S\+'))
+  " remove all those backspaces
+  silent execute 'keepjumps %substitute,.\b,,ge'
```
I guess you need `:keeppatterns` here and elsewhere where substitute is used.

justinmk replied:
Also why is `execute` used? Doesn't seem needed.

nhooyr replied:
Will fix, not quite sure why execute is there actually. I think its an artifact from something funky I was doing when experimenting with the best way to handle the backspaced text.

mhinz commented on runtime/autoload/man.vim:
```diff
@@ -112,31 +112,18 @@ function! s:parse_page_and_sect_path(path) abort
-  endwhile
-  keepjumps 1
+  silent execute 'keepjumps %substitute,.\b,,ge'
+  keepjumps 1delete _
```
Needs a `:silent`.

nhooyr replied:
will fix.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -11,127 +20,184 @@ catch /E145:/
+  " TODO(nhooyr) perhaps do not load, merely redisplay?
+  " remove all the text, incase we already loaded the manpage before
+  keepjumps %delete _
+  let $MANWIDTH = winwidth(0)-1
```
@justinmk do you know a better solution to this? 

http://stackoverflow.com/questions/26315925/get-usable-window-width-in-vim-script

I want 3 but I'm getting 1 here.

nhooyr commented:
Ooh. Nevermind, we got a small problem. Because we calculate `$MANWIDTH` and then read the manpage in, if the user has an auto command that turns on `number` or `relativenumber` or something else that causes a decrease in the number of columns, `$MANWIDTH` would have been calculated incorrectly. Thus I think we should just not bother with calculations and just use `winwidth(0)`.

@justinmk thoughts?

nhooyr commented:
Maybe we could have an auto command that the user can hook into to set window local options in a new window?

nhooyr commented:
Ah never mind, I just removed it. I think `winwidth(0)` will be perfectly adequate for everyone involved. I also removed the usage of `keywordprg` because we are not setting `iskeyword` and thus it does not work correctly. 

The manual mappings work better imo because they do not rely on some preset characters that can appear in the name of a manpage. They simply let `man` decide whether it is a valid manpage or not. We also avoid all of the `iskeyword` stuff.

I've fixed a bunch of stuff as well. I'll push tomorrow, gotta sleep now.

nhooyr commented on runtime/autoload/man.vim:
```diff
@@ -1,6 +1,15 @@
+let s:tag_stack = []
 
 try
   if !has('win32') && $OSTYPE !~? 'cygwin\|linux' && system('uname -s') =~? 'SunOS' && system('uname -r') =~? '^5'
```
@justinmk I'm removing this because its a fix for only Solaris/SunOS and we do not support them.

justinmk replied:
@nhooyr We support illumous, which is a variant of Solaris. Not sure if it's relevant to this though.

justinmk replied:
libuv even seems to support solaris, so that means we support it: https://github.com/libuv/libuv/commit/b27a55c3cc4b8a80b4713d0dfa94a164ff430487

nhooyr replied:
oh ok, just that here it says we do not: https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim

nhooyr replied:
so then I should just change SunOS there to illumous? The [flags](https://illumos.org/man/1/man) look right

justinmk replied:
> here it says we do not

Thanks, fixed. That was written in a time when libuv was younger.

> so then I should just change SunOS there to illumous

Sorry, I don't know. The least-risky thing is to do whatever Vim's man.vim does, unless there's an explicit reason not to.

nhooyr replied:
It does this exactly. I'll leave it alone.

nhooyr commented:
Ok, so I have a new more elegant solution (perhaps could be better, but I'm satisfied) to the default count problem. I removed `g:man_find_window` because its just extremely unlikely that someone will ever want to open a new tab every single `:Man`. I've fixed quite a few bugs and made the regexes much better.

I'm fairly certain I'm done now. I've looked over the code several times, tested all of it repeatedly and it all looks good. It's ready for merging.

Since we're sorting 
TODOS (maybe) :
- [ ] I do not think it will work on Solaris because I'm not sure how to get `man --path` there (not `$MANPATH`) for the autocomplete.
- [ ] Since we're sorting candidates and formatting candidates better, `man#complete` may need optimization.
- [ ] Backspaced syntax highlighting.
- [ ] How to handle splitting in multiple directions.
- [ ] A fix for the autocomplete bug (still gotta report it to vim).
- [ ] Perhaps a separate ftplugin/command for when using neovim as a pager? See [vim](https://github.com/vim/vim/blob/06d2d38ab7564e1f784b1058a4ef4580cd6d1810/runtime/plugin/manpager.vim)

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,254 @@
-      let sect = a:sect
+call s:init_mandirs()
+
+" TODO(nhooyr) maybe optimize?
```
too many TODOs.

nhooyr replied:
fixed, only 4 now.

justinmk commented:
The TODOs don't seem important to me. If you decide to send PRs for them, please make them small and focused :)

If this is ready, change title to RFC. Thanks for following through with this!

nhooyr commented:
Alright, its down to two TODOs. One for testing on Solaris and the other for the autocompletion bug. Yes, any follow up PRs will be small and focused.

No problem, I enjoyed finishing it. Felt bad to just leave it hanging for so long.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,252 @@
+    " then call it every time the buffer is loaded so that it is always set
+    " correctly for man.vim but whatever it was before for every other
+    " filetype; by default.
+    setlocal filetype=man
```
In gitter we discussed setting window-local options _only_ if man.vim creates a new window or tab. It must not set window-local options if a new window/tab was not created.

nhooyr replied:
fixed

justinmk commented:
Tested this briefly. 

Problem I noticed:

```
edit foo.txt
:Man ls  " opens a new tab; good
:tabnew
:edit bar.txt
:Man printf " loads manpage in the exist window; not good
```

`:Man` should never load in a non-manpage window. Find the existing manpage tab, I guess, and if it doesn't exist, create a new one.

It seems like many changes in this PR were made with the intent of using vim as a manpager. That's going to annoy people (like me) who just want to use `:Man`. Any change that favors using vim as a manpager and _also_ regresses the old `:Man` behavior should not be made.

nhooyr commented:
Fixed the problem, except for searching tabs. The old `man.vim` only searched through open windows, not tabs. Are you sure you want tab searching as well?

nhooyr commented:
By the way, I never intended to make any change that favoured using vim as a manpager. I myself just use `:Man`. I was hoping to add support for anyone who did want to use vim as a manpager. That was just a small bug.

tweekmonster commented:
> Are you sure you want tab searching as well?

I would say search for a `man` window in the current tab.  If there isn't one, find a tab that only has one `man` window, then display the tab and replace the window.  If there isn't one, create a new tab regardless of another tab with multiple windows that has `man` among them.

My reasoning is that if the user already has a `man` page in view, they would expect to see it replaced there.  If not, it'd be easy to understand a new tab being switch to if it only has one window.  On the other hand, switching to another tab with multiple windows might give you pause if it seems like the layout or number of windows has changed.

justinmk commented:
@tweekmonster's description is what I had in mind, but I only suggested it as a compromise if you wanted to go that direction. Can be a separate PR or just left out.

justinmk commented:
Found a regression:

 On linux, do e.g. `:Man printf(3)`. Find the text that looks like this (just an example, there are many cases):

       snprintf(), vsnprintf():
           _BSD_SOURCE || _XOPEN_SOURCE >= 500 || _ISOC99_SOURCE || _POSIX_C_SOURCE >= 200112L;
           or cc -std=c99

`K` and `ctrl-]` on `snprintf(),` fails with:

    man.vim: invalid manpage reference snprintf(),

Similar for `vsnprintf():`.

mhinz commented:
- Now I have an empty line at the top again with `:Man printf(3)` and `man 3 printf`.
- If there are code examples, some functions are marked as manReferences and some are not.
- From the marked ones, a few a false positives, e.g. `sizeof(buffer)` can't take me to any manpage, since `sizeof` is an operator. `K` on it takes me to `buffer(3)` instead.

nhooyr commented:
@justinmk That was intended, as `vsprintf()` that isn't a valid manpage reference. A manpage reference with parenthesis in it always has a section in between the `()`. It's not even highlighted as a manpage reference. I can still make it work with `K` or `<C-]>` if you really want.

In regards to @tweekmonster's description, I'll do that in a follow up PR.

@mhinz 
1. fixed.
2. I've made the `manReference` highlighting more specific.
3. same regex as above was fixed.
4. same problem as 1, so fixed (empty line at the top meant an top filename was read).

nhooyr commented:
@justinmk

Try
```
:set number
:Man printf
:Man ls
```

See how there are now numbers on the side? In order to fix this, I also have to set window local options when we are already on, or have found the closed `man` window. That is how the old `man.vim` does it. I also noticed a bug in the old `man.vim`.

Try
```
:set number
:Man printf
:vnew
:Man ls
```

See numbers on the side? I just have to add one call to fix these problems but I want to confirm with you.

justinmk commented:
> That was intended, as vsprintf() that isn't a valid manpage reference

It worked in the old plugin. The fact that it gets confused by the parens now doesn't seem like an improvement. What was gained?

> By the way, :help does not have any of these problems.

`:help` always opens a new window, except when following a tag in a help window. 

At this point I would be in favor of having the same behavior as `:help`. Those who want a new tab can use `:tab Man foo` (depends on resolution of https://github.com/neovim/neovim/issues/4751#issuecomment-219191833 which we don't need to address here, it requires an un-merged Vim patch).

By the way, I just realized that with the old plugin, `:Man foo` on an existing `man://` window re-uses that window. So if it makes things easier, that behavior is fine (I objected to it yesterday). Sorry for the confusion.

nhooyr commented:
@justinmk 
I fixed the paren issue. 

`:help` does not always open a new window.

Try
```
:help :command
:vnew
:help help
```

`:help` uses the existing window in the current tab, and only if there is none, it creates a split.

Ok so I'll make it behave exactly like `:help`.

nhooyr commented:
Ok, it should behave just like `:help` now.

nhooyr commented:
When using `buf` to switch to an already open manpage, a message like this appears: `"man://echo(1)" [readonly] line 1 of 30 --3%-- col 1`

Should it be silenced?

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,264 @@
-  silent keepjumps norm! 1G"_dG
-  if empty($MANWIDTH)
+  setlocal noreadonly
+  if empty($MANWIDTH) || $MANWIDTH > winwidth(0)
```
This will partially regress https://github.com/neovim/neovim/pull/4144. The reason for respecting MANWIDTH even if it is wider than the current window is that the current window might only be temporarily narrow. Since we don't reflow, we should just assume the user knows what they're doing and _actually respect_ MANWIDTH. (i.e., undo the change on this line)

nhooyr replied:
fixed

nhooyr commented:
Found a couple edge cases. Fixing them.

nhooyr commented:
@justinmk: `:help` always jumps to the top, even if the buffer was opened before. Just in interest of making it behave more like `:help` should we do that here as well?

Or should I just add an option but make the default always jump to the top?

justinmk commented:
Up to you. We can change it if there are complaints.

nhooyr commented:
Ok, it moves to the top now because I think that is what most people will expect it to do.

justinmk commented:
At `:Man printf(3)`, find this line:

    if ((p = malloc(size)) == NULL)

`K` on malloc fails.

nhooyr commented:
Same issue as above, since it's technically an invalid manpage, I intentionally did not allow it. But it's permitted now.

nhooyr commented on runtime/doc/filetype.txt:
```diff
@@ -512,37 +512,44 @@ Local mappings:
+                        If set, manpages are folded with foldmethod=indent and
+                        foldnestmax=1.
 
+If you do not like the default folding, use an autocommand to add your desired
```
Is this really necessary? I do not think anyone who uses vim (or even got that far into the docs) needs to know how to use `:autocmd` to set their favourite folding style.

nhooyr replied:
Not sure why this is an outdated diff, but whatever. Question still applies.

nhooyr commented on runtime/ftplugin/man.vim:
```diff
@@ -1,41 +1,53 @@
+setlocal shiftwidth=8
+call man#set_window_local_options()
+
+if !exists('g:no_plugin_maps') && !exists('g:no_man_maps')
```
@justinmk I find it odd that here we only check if these options exist, but below, we check if `ft_man_folding_enable` is set to 1. Should I leave it?

justinmk replied:
For better or worse that is the convention for these particular variables, though I don't think new variables should follow it. E.g. g:no_plugin_maps is checked in the same way elsewhere, use grep to find other cases.

nhooyr replied:
Yea, its just convention. the `no_maps` variables are always checked to see if they exist, but other variables are always checked against 1. Weird convention.

Btw, because the diff is outdated, did you not get a notification for https://github.com/neovim/neovim/pull/4449#discussion_r72910313

nhooyr commented:
Aside from my two questions on the diffs, I think this is finally done. I know I've said it like 5 times before, but it really looks done now!

🎉

nhooyr commented:
oh dammit. I noticed a new problem. Since we assign to `$MANWIDTH` and then since it becomes non-empty, if the window size changes, `$MANWIDTH` still says the same and never adjusts.

Gonna fix.

nhooyr commented:
Ok so now we do not directly assign to $MANWIDTH. Rather, we use a local variable. This way, we can check if the user explicitly set $MANWIDTH, and if so we obey. Otherwise, we can freely adapt to any size because for neovim, $MANWIDTH will not be set.

nhooyr commented:
Same problem with using `:buf` to switch to an already loaded manpage. It won't always fit the current window. I've taken out that functionality for now. I'll add it back in a later PR where I check whether the original `$MANWIDTH` it was opened with has changed, and if so, only then do I reload. That way, we respect `$MANWIDTH` if set, but if not, it opening the page again will always resize it if needed.

But otherwise I've squashed and I added/updated the maintainer headers. So done!

mhinz commented:
Is this RDY?

justinmk commented:
creating a new buffer every time does not sit right with me I'd rather leave that in and fix the *other* problem  later

jamessan commented:
> So now, if $MANWIDTH is not set, we do not assign directly to $MANWIDTH because then $MANWIDTH will always stay set to the same value

Why?  I recently fixed this issue in Vim's man.vim with this patch:

```diff
@@ -160,10 +170,15 @@ func <SID>GetPage(...)
 
   setl ma nonu nornu nofen
   silent exec "norm 1GdG"
+  let unsetwidth = 0
   if empty($MANWIDTH)
     let $MANWIDTH = winwidth(0)
+    let unsetwidth = 1
   endif
   silent exec "r!/usr/bin/man ".s:GetCmdArg(sect, page)." | col -b"
+  if unsetwidth
+    let $MANWIDTH = ''
+  endif
   " Remove blank lines from top and bottom.
   while getline(1) =~ '^\s*$'
     silent keepj norm ggdd
```

You could do something similar here.  If not, then I'd suggest using `env` instead of prefixing the command with the variable.  That doesn't work in all shells.

    env MANWIDTH=80 man ...

nhooyr commented:
@justinmk It'll be the same. Not sure what you mean. I still reuse the buffers, but now, I delete everything in them if it was already loaded and then reload the page. See [here](https://github.com/nhooyr/neovim/blob/90e6015431fe6582f318ee05ea9dc75bc8af5041/runtime/autoload/man.vim#L154). Earlier, I just switched to the buffer and set some options. Since I didn't reload the page from `man`, the width was not guaranteed to fit. Now I reload it from `man` always (the downside to this is that jumps become invalid).

What I'll add in another PR is reloading from `man` only if the width that the page was previously opened at has changed. I can add it right now if you want.

@jamessan what shell does prefixing environment variables not work in?

jamessan commented:
At least fish, and possibly others.  `env` is the canonical way to adjust the environment for a single command.

nhooyr commented:
Alright, will fix.

nhooyr commented:
Ok so I'm using `env` now to modify the environment and I just went ahead and added automatic reloading. It was only 15 more lines. If the manpage was already opened but if it were to be reloaded, if the width at which it would be opened is the same as the width at which it was opened previously, then instead of reloading from `man`, we just switch to it. Otherwise we reload.

justinmk commented on runtime/autoload/man.vim:
```diff
@@ -1,137 +1,290 @@
+  return [sect, name]
+endfunction
+
+let s:manwidths = {}
```
Why not use a `b:` variable (can be retrieved via `getbufvar()`) instead of re-inventing the `b:` dictionary here?

nhooyr replied:
I forgot about buffer local variables. will fix.

nhooyr commented:
fixed, using a buffer local variable now.

tweekmonster commented:
I made a small addition that uses the location list as a TOC which I find super helpful:  https://github.com/tweekmonster/neovim/commit/94a4b316516b92dcb8938b8b557c645abc9be79b

Wanted to get everyone's thoughts on it being added.  I can submit it as a separate PR after this one is merged to keep the current goals clear.

nhooyr commented:
@tweekmonster Is it better than remapping `[[` and `]]` to do the same thing? Like how it works in [vim-man](https://github.com/vim-utils/vim-man)

tweekmonster commented:
I think it's better if I know the section that I want to get to or want an overview of the sections.

I found myself wanting to get to the `FORMATS` section in the tmux manual a lot recently.  Searching `/FORMATS` is annoying because it mentions "see FORMATS" more than a couple times.

Now, it's just `:lwindow` and `/FORMATS`.

justinmk commented:
Nice work. The auto-completion works well.

@tweekmonster `:lwindow` for TOC is a great idea. Could be used for help windows too, and could remove the need for the ugly TOC at the top of many helpfiles...

tweekmonster commented:
Can't believe I didn't think of the docs, too.  I'll put together a PR.

KillTheMule commented:
@nhooyr Great! I'd like to put a post on reddit for this feature, but I never used man.vim and I can't check it out now (computer broke). Could you give 2-3 or so real-world examples for the top improvements of this over the old man plugin?

nhooyr commented:
@KillTheMule I'm gonna post it myself. I'm just waiting until @tweekmonster's PR for TOC is merged in.

KillTheMule commented:
Awesome :+1:

oni-link commented:
I like to use a count with `K` to see a specific section, but this does not work anymore:
With `2K` on word `exit` I only see the error: `E481: No range allowed: 2:Man exit`

Is this a bug or do we have to use `:Man 2 exit` now?

nhooyr commented:
@oni-link Unset `keywordprg` and make your own mapping as so:

```vim
nmap <silent> K <Plug>(Man)
```

nhooyr commented:
@justinmk I completely forgot, we have to do something about the default of `keywordprg` (the new plugin does not work with `keywordprg` because it forces the use of `iskeyword`). We should make it empty by default and set a default global mapping for `K` to `<Plug>(Man)`.

oni-link commented:
@nhooyr, thanks for the mapping. The help page says using `<Plug>(Man)` would open a new tab, but it opens only a window.

nhooyr commented:
@oni-link good catch, the help page needs to be fixed.

jamessan commented:
> keywordprg because you cannot check if the user explicitly set a section

Why not?  Can't you use `v:count` to get that?

nhooyr commented:
@jamessan https://github.com/neovim/neovim/pull/4449#discussion_r71965893

jamessan commented:
Ah, thanks.  Forgot about that part of the discussion. :)

nhooyr commented:
Actually, I was wrong. I was under the impression that `v:count1` is only available for mappings for some reason. Turns out, it works for commands! I can fix it in #5168

nhooyr commented:
@oni-link If you update, you will not need the mapping anymore!

oni-link commented:
@nhooyr, there seems to be an inconsistency with section use in `:Man`.
`:Man 7 pipe` shows section `2` for `pipe`.  `:Man pipe(7)` and `7K` show section `7`.

nhooyr commented:
@oni-link thanks, you identified another problem. But try and make new issues instead of reviving this one.

This is a weird problem. I've opened #5202.

dumblob commented:
I'm glad man pages are now these "native first class citizens", but there is still one issue remaining (I'm sorry for not reporting it earlier). Namely the hardcoded **horizontal** splitting (`execute 'split '.bufname`). Can you think of any easy way how an end user could change it to **vertical** splitting?

By the way there are even [patches](https://www.reddit.com/r/vim/comments/2irn8j/vertical_split_by_default/ ) for vim trying to change globally the split-horizontally-by-default rule to split-vertically-by-default, but they didn't make it to neovim as far as I can tell. Might be worth thinking about making the default splitting behavior configurable in some of the neovim future releases as the width of current displays grows unlike their height.

nhooyr commented:
@dumblob that was fixed in a later PR, you can just do `vertical Man printf`, or `tab Man printf` or even any other command modifier.

dumblob commented:
@nhooyr thanks for the advice. But how to do make it default for `3K` which I use the most.

nhooyr commented:
@dumblob you can just map it to K

`nnoremap <silent> <buffer> K          :vertical Man<CR>`

nhooyr commented:
Make sure you set `g:no_man_maps` first and you'll have to remap all the keys. Look at https://github.com/neovim/neovim/blob/40c76741c187f5bf35101e65252226030d5b72e5/runtime/ftplugin/man.vim#L33-L42

Just copy it, add vertical and then put it into `$XDG_CONFIG_HOME/nvim/after/ftplugin/man.vim`. And then enjoy :)

dumblob commented:
Hi @nhooyr, thanks for the pointer, but this unfortunately doesn't work with ranges (`E481: No range allowed`). Any other ideas (I didn't try to set `g:no_man_maps` and remap all keys as I still believe there is a better solution)? If not, I'll open a normal issue instead of commenting on this PR.

Shougo commented:
I think `:Man` must support the range.

justinmk commented:
Don't remap K, just set keywordprg.

    :set keywordprg=:vertical\ Man

Shougo commented:
OK.

nhooyr commented:
@justinmk that doesn't work with a range.

nhooyr commented:
afaict, you have to do this:

```
command! -range=0 -complete=customlist,man#complete -nargs=* Vman call man#open_page(v:count, v:count1, 'vertical', <f-args>)
set keywordprg=:Vman
autocmd FileType man nnoremap <silent> <buffer> K :Vman<CR>
```

nhooyr commented:
@justinmk `:Man` does support range! `:vertical` doesn't support a range. perhaps it should?

justinmk commented:
@nhooyr I recall now that we decided not to support range because `v:count` was used to make some decision (related to behavior of `q`, maybe). Are we still doing that? Perhaps supporting range for `:Man` is more important than `q`.

dumblob commented:
>afaict, you have to do this:
>...

Thanks @nhooyr, it works! I wouldn't figure this out myself.

nhooyr commented:
@dumblob 

You can also do

```
nnoremap <silent> K :<C-u>execute 'vertical' (v:count ==# v:count1 ? v:count : '') 'Man'<CR>
autocmd! FileType man nnoremap <buffer> <silent> K :<C-u>execute 'vertical' (v:count ==# v:count1 ? v:count : '') 'Man'<CR>
```

This way, you don't have to create a new command. Though the downside here is that you can't set `'keywordprg'`, though I don't think that has any advantages.
